src/runtime/vm/vm.cc

```diff
@@ -595,6 +601,29 @@ void VirtualMachine::RunLoop() {
         pc_++;
         goto main_loop;
       }
+      case Opcode::RefRead: {
+        ADT ref = Downcast<ADT>(ReadRegister(instr.ref_read.ref));
+        WriteRegister(instr.dst, ref[0]);
+        pc_++;
+        goto main_loop;
+      }
+      case Opcode::RefWrite: {
+        ADT ref = Downcast<ADT>(ReadRegister(instr.ref_write.ref));
+        ObjectRef value = ReadRegister(instr.ref_write.value);
+        // Not sure about this being best way to implement mutable thing.
+        ADTObj* mut_array = const_cast<ADTObj*>(ref.as<ADTObj>());
+        ObjectRef& inner_value = (*mut_array)[0];
+        inner_value = value;
+        pc_++;
+        goto main_loop;
+      }
+      case Opcode::RefCreate: {
+        auto value = ReadRegister(instr.ref_create.initial_value);
+        auto ref = ADT(111, {value});
```

Comment:
How we do ensure this tag does not collide with other ADT's tag number?

Reply:
It doesn't actually matter, I just used ADT container since typing guarantees we already know ahead of time the value is a reference we don't need to check or inspect the tag.

Reply:
Maybe in this case `Tuple({values})` is okay which defaults tag to 0?

Reply:
Let me speculate a case that in a relay program we defined many ADTs (150 ADT types), and VM compiler generates tags for each ADT type, is it possible to collide with 111?  I recall it includes hashes in the tag number, so probably unlikely. And `Tuple` seems perfect fix.